apps/src/dance/danceRedux.ts

```diff
@@ -145,10 +145,9 @@ async function handleSongSelection(
   // manifest within Dance.js, and Lab2 Dance which reads the current song's manifest from Redux.
   if (onSongSelected) {
     onSongSelected(songId);
-  } else {
-    const metadata = await loadSongMetadata(songId);
-    dispatch(setCurrentSongMetadata(metadata));
   }
+  const metadata = await loadSongMetadata(songId);
+  dispatch(setCurrentSongMetadata(metadata));
```

Comment:
Curious for thoughts on this - this change means that we'll fetch song metadata twice instead of just once, for every song selected. In practice this shouldn't be a huge issue since the browser should cache the file, but something note.

The reason for making this change is that we need to grab the song metadata from redux in `AiVisualizationPreview`. Previously, the thought here was that we would always either just fetch the song metadata in Dance.js directly (legacy dance) or only fetch it in Redux (lab2 dance), but now we're doing a combination of both.

Reply:
That seems reasonable to me!

Reply:
If you use the same song metadata for animated preview, then could you circumvent this?

Reply:
I was thinking about this a bit for static preview (I haven't added the text background preview yet), and was planning on building on this to use song names / artist names for it.

I think having real song names / artists will be nice for the static previews -- if we don't have them, I think we'd need to use english-only words in the preview, or pass through a translation of the word "preview" (at which point, might as well just pass the real song names / artist).

If we need specific values for song metadata (centroid, etc), maybe we hard code that info but keep using the selected artist / song name?

Reply:
> I was thinking about this a bit for static preview (I haven't added the text background preview yet), and was planning on building on this to use song names / artist names for it.
> 
> I think having real song names / artists will be nice for the static previews -- if we don't have them, I think we'd need to use english-only words in the preview, or pass through a translation of the word "preview" (at which point, might as well just pass the real song names / artist).
> 
> If we need specific values for song metadata (centroid, etc), maybe we hard code that info but keep using the selected artist / song name?

Oh - cool idea! So have sample song metadata for all songs but retrieve the song name in case the 'text' effect is selected? Is there another case we need the selected artist/song name in the preview?

Reply:
I was thinking continue to do what Sanchit is doing here (pass in whole selected song metadata via `this.nativeAPI.livePreview(songMetadata);`), but override any values in `songMetadata` (eg, centroid) that cause problems when rendering previews.

Reply:
Yeah, the "text" effect is mainly what makes this tricky - otherwise having sample metadata would work pretty well in that we can tailor to work for all effects, and to your point Alice, yeah we wouldn't need to fetch metadata twice here. That being said, there aren't too many effects that need to be handled, and the solution I was thinking was simply setting a lower bound for the centroid value (whatever the static preview value is) so they always showed up

Reply:
> Yeah, the "text" effect is mainly what makes this tricky - otherwise having sample metadata would work pretty well in that we can tailor to work for all effects, and to your point Alice, yeah we wouldn't need to fetch metadata twice here. That being said, there aren't too many effects that need to be handled, and the solution I was thinking was simply setting a lower bound for the centroid value (whatever the static preview value is) so they always showed up

That sounds good. We have so many customizations at this point...hopefully we'll get a chance to refactor the `Effects.js` file in `dance-party` at some point.

Just throwing out an idea, what do you think of reassigning `songMetadata` in `ProgramExecutor.ts` [here](https://github.com/code-dot-org/code-dot-org/blob/2f828b15937d00c49a6937b048d7006ce1ed6670/apps/src/dance/lab2/ProgramExecutor.ts#L123) with sample data but the title and artist name be maintained?

Reply:
> Just throwing out an idea, what do you think of reassigning songMetadata in ProgramExecutor.ts [here](https://github.com/code-dot-org/code-dot-org/blob/2f828b15937d00c49a6937b048d7006ce1ed6670/apps/src/dance/lab2/ProgramExecutor.ts#L123) with sample data but the title and artist name be maintained?

Yep! we could do some transformation on the metadata, or even do that within the native layer when calling `livePreview()` - I think for now to get this PR through I'll stick with this approach but let's discuss as a follow up